### ISSUE TITLE
Release binaries for desktop-ish Linux: how? AppImage/Flatpak/something else?

### ISSUE THREAD

**Issue by ghost** [hikari-no-yume/touchHLE#77]
Linux AppImage/Flatpak support for this would be neat.

**Comment by upintheairsheep** [comment]
Does not support Linux, please use wine.

**Comment by ghost** [comment]
Will close for now, but please think about adding Linux support in the future.

**Comment by hikari-no-yume** [comment]
Uh, to be clear, this project does “support” Linux, in the sense that you can build it for Linux, it's just not something we promise will be easy or work well, for the time being at least.

There was an existing discussion about Linux support in https://github.com/hikari-no-yume/touchHLE/issues/3.

I don't really want to do Linux binary releases, it's not a single platform in the way that Windows or macOS are. But I am willing to make it easier to build yourself on Linux.

**Comment by hikari-no-yume** [comment]
Also, if you hit any problems while building it yourself, please tell us about them (probably in the thread I linked).

**Comment by ghost** [comment]
Thanks for the clarification

**Comment by WingofaGriffin** [comment]
I want to revive this discussion a bit.

It seems that the biggest objection to Linux binaries being released is due to all the different distros and dependencies that differ. I would argue that Flatpak and AppImage seek to solve that issue directly, by specifically being distro agnostic.

Literally from the [AppImage front page](https://appimage.org):
> "As a user, I want to download an application from the original author, and run it on my Linux desktop system just like I would do with a Windows or Mac application."
>"As an application author, I want to provide packages for Linux desktop systems, without the need to get it 'into' a distribution and without having to build for gazillions of different distributions."

These two package types bundle the deps in with the binary (a bit more complicated than that, but you get the jist). When paired with a CI tool like GitHub actions, which you already seem to be using, it should be pretty easy to automate a new set of builds, especially with the GitHub provided Ubuntu instance.

Let me know if you want to discuss further, I appriciate the existing Linux support you do have, and have given, but want to see if we can make this easier for all of us :)

**Comment by LunaTheFoxgirl** [comment]
I'd recommend going with flatpak if anything for touchHLE, AppImage's main developer can be a living nightmare to deal with, he has a tendency to harass other devs, drop AppImage PRs without any will to maintain them. + Flatpak has the upside of being a stable environment to have your app in and does de-duplication of dependencies.

**Comment by hikari-no-yume** [comment]
Ah, thanks for the heads up.

Unrelated: I wonder if I can avoid using any of these imaging formats. I've refused to have “Linux” builds because it doesn't seem like a well-defined platform to me, but actually, touchHLE probably only needs glibc, SDL 2, and maybe a handful of other things. That's a small enough surface area that I think imaging/containerisation/etc is superfluous. I should do some experiments.

**Comment by LunaTheFoxgirl** [comment]
If you build on something like debian, then yeah it might just work, both glibc and SDL2 has backwards compatibility built in

**Comment by hikari-no-yume** [comment]
Maybe I should look into what the Steam runtime provides. touchHLE is very similar to a typical game in terms of what stuff it needs, and it's probably all in there. If there are distros where Steam support is unrealistic, I would say binaries for touchHLE are also unrealistic there.

**Comment by LunaTheFoxgirl** [comment]
> Maybe I should look into what the Steam runtime provides. touchHLE is very similar to a typical game in terms of what stuff it needs, and it's probably all in there. If there are distros where Steam support is unrealistic, I would say binaries for touchHLE are also unrealistic there.

Steam runtime provides basically a old ubuntu container, last time I checked it was 18.04

**Comment by hikari-no-yume** [comment]
If I stick to SDL2, glibc and libstdc++, say, should it work also on stuff that's not Debian-derived, assuming I use reasonably old minimum versions?

**Comment by LunaTheFoxgirl** [comment]
More or less, once you get UI in to touchHLE I would recommend releasing it as a flatpak, means it's one package that just works on linux systems, even non-glibc ones as flatpak ships its containers with it that anyone can download on virtually any distro.

**Comment by ghost** [comment]
A flatpak would be amazing!
flatpak is a single platform that is platform agnostic, and is much like Android or macOS!
it solves many issues and nightmares with Linux both for developers and users, and is generally amazing!
plus, a flatpak release would mean compatibility with the Steam Deck too!
you can ship it to flathub or as a .flatpak, whichever you prefer

**Comment by puigru** [comment]
Agreed, flatpak support would be great

**Comment by Twig6943** [comment]
Whilst the devs are waiting upon making a GUI for the app first (see https://github.com/touchHLE/touchHLE/issues/3#issuecomment-1423806750) I'll try start doing some packaging here soon https://github.com/Twig6943/touchhle-packaging PRS are welcome

( @DCNick3 tell me if you have some leftover stuff)

**Comment by hikari-no-yume** [comment]
I can't stop you, but please don't use a namespace that doesn't belong to you (`touchhle.github.io`).

**Comment by Twig6943** [comment]
> I can't stop you, but please don't use a namespace that doesn't belong to you (`touchhle.github.io`).

It's a placeholder, I won't actually publish it unless you tell me to

**Comment by Twig6943** [comment]
@Samueru-sama has packaged the app as an AppImage. Please give feedback https://github.com/pkgforge-dev/TouchHLE-AppImage

**Comment by hikari-no-yume** [comment]
That one seems to rely on some touchHLE packaging effort in AUR by yet another person.

**Comment by Samueru-sama** [comment]
> That one seems to rely on some touchHLE packaging effort in AUR by yet another person.

yeah I took the [AUR package](https://github.com/pkgforge-dev/TouchHLE-AppImage/blob/17fc36ebf4c44efc54c7abeba193280ffd4ad435/get-dependencies.sh#L40-L47) to build the binary.

Btw reading the other comments here, you don't have to build the binary on old systems to target linux, if you use [sharun](https://github.com/VHSgunzo/sharun) it makes them able to work on any linux system, here it is running on alpine linux (no glibc distro).


<img width="800" height="797" alt="Image" src="https://github.com/user-attachments/assets/4f20fc0f-316b-4342-8b1f-e1f1730ae3d5" />

**Comment by Twig6943** [comment]
> That one seems to rely on some touchHLE packaging effort in AUR by yet another person.

I emailed the person behind the PKGBUILD about it. I'm 99% sure they're okay with it but will still report what he says.

**Comment by hikari-no-yume** [comment]
Having read a bit of the AppImage documentation I'm not really convinced of the quality of that particular system. Since an AppImage is just an executable anyway, maybe we could just ship a static build of touchHLE if people really want that. Flatpak looks better-designed, though the particular way it handles GPU drivers is quite unfortunate.

To be honest, though, I don't really want to maintain Linux packaging, and I question what the actual benefit to users would be. touchHLE works well in WINE and I can recommend running it that way.

**Comment by 27hectormanuel** [comment]
I'd like to use it on Steam deck

**Comment by Samueru-sama** [comment]
> Having read a bit of the AppImage documentation I'm not really convinced of the quality of that particular system.

Read this instead: https://github.com/pkgforge-dev/Anylinux-AppImages/blob/main/HOW-TO-MAKE-THESE.md

The AppImage documentation is abandoned and says a lot of things that are not true anymore, it does not even make mention of [Go-appimage](https://github.com/probonopd/go-appimage) which works similar to sharun.

See also https://github.com/AppImage/docs.appimage.org/pull/102

EDIT: This specific AppImage even uses a runtime [that does not need FUSE at all](https://github.com/VHSgunzo/uruntime) to work

>maybe we could just ship a static build of touchHLE if people really want that. 

Sure, if you can make a fully static binary that works better than AppImage, however that means: 

* You will have to link mesa statically (in this AppImage I ship a smaller version that makes the final AppImage **less than 30 MiB**). 

* You will not have hardware acceleration with nvidia, because the proprietary nvidia driver is massive and cannot be bundled due to its license.

**Comment by Twig6943** [comment]
> > That one seems to rely on some touchHLE packaging effort in AUR by yet another person.
> 
> I emailed the person behind the PKGBUILD about it. I'm 99% sure they're okay with it but will still report what he says.

He told me he's okay with it.

**Comment by hikari-no-yume** [comment]
The whole concept of bundling your own graphics driver (Mesa in this case) into a Flatpak runtime or AppImage is absurd to me. This is why people don't take Linux seriously! Every 3D graphics API is designed with a stable ABI for dynamic linking, so that you aren't locked into a particular vendor for your hardware or drivers. Why would you sacrifice that willingly… it's a great way to ensure that the app won't work with next year's hardware.

**Comment by Samueru-sama** [comment]
> so that you aren't locked into a particular vendor for your hardware or drivers.

>  it's a great way to ensure that the app won't work with next year's hardware.

**You are not locked into a specific version with sharun!** You also don't need to include mesa drivers either, however that means that hardware acceleration is broken in non-glibc systems. 

I used to not ship mesa, but then got these issues: 

* https://github.com/pkgforge-dev/mpv-AppImage/issues/19

* https://github.com/pkgforge-dev/Citron-AppImage/issues/14

So now using the host mesa is locked behind an env variable in sharun.

By default with `aarch64` we still allow using the host mesa drivers, because those often have vendor specific patches that have not been upstreamed. 

**EDIT:** [Also with nvidia we always use the nvidia driver of the host.](https://github.com/pkgforge-dev/Anylinux-AppImages/blob/main/HOW-TO-MAKE-THESE.md#what-about-nvidia)

**Comment by nighto** [comment]
Seems to be running fine on Ubuntu:

<img width="1042" height="629" alt="Image" src="https://github.com/user-attachments/assets/31a5a7d0-d627-4543-827b-a78934ce3935" />

<img width="1042" height="629" alt="Image" src="https://github.com/user-attachments/assets/efcd1b71-8679-4ed1-b218-5f9e669ee9f3" />

**Comment by nighto** [comment]
<img width="668" height="365" alt="Image" src="https://github.com/user-attachments/assets/3d82c8cf-2448-49f6-bef2-6d3966c5d29b" />

btw this keeps popping up even after pressing yes to allow, appeared maybe 5 times, including 1 time after closing the app

**Comment by Samueru-sama** [comment]
> btw this keeps popping up even after closing the app

That question will show up if you agreed to check for updates earlier, [the updater needs appimageupdatetool to perform updates](https://github.com/pkgforge-dev/Anylinux-AppImages/blob/main/useful-tools/hooks/self-updater.bg.hook#L78).

maybe you don't have `wget` or `curl` installed? that's the only thing that would explain asking again if you said yes to download.

**Comment by nighto** [comment]
I have both `wget` and `curl` installed, just checked.

Just tested opening again and it didn't ask this time, so maybe it was just the first time?

**Comment by Samueru-sama** [comment]
> I have both `wget` and `curl` installed, just checked.

weird, the script will also aks to disable the [update checking if it fails to download appimageupdatetool](https://github.com/pkgforge-dev/Anylinux-AppImages/blob/main/useful-tools/hooks/self-updater.bg.hook#L142-L144)

>Just tested opening again and it didn't ask this time, so maybe it was just the first time?

Maybe what happened is that it could not download appimageupdatetool during that time? check if you have it now in `~/.local/bin`


**EDIT:** The script already notifies when both wget and curl are not available, so the only explanation I have is that the download was failing, which would trigger the `set -e` and cause the script to stop without asking to disable updates, will add a check for that.

**Comment by nighto** [comment]
I do have a `~/.local/bin/appimageupdatetool` executable file (amongs other unrelated files).

**Comment by Samueru-sama** [comment]
> I do have a `~/.local/bin/appimageupdatetool` executable file (amongs other unrelated files).

Yeah most likely what happened was that it was failing to download for some reason, this would cause the script to exit quietly since it has `set -e`.

[Just added this to avoid that happening in the future.](https://github.com/pkgforge-dev/Anylinux-AppImages/commit/7419637c83d30df44f9846b680c53cb888e7773b)

**Comment by Link4Electronics** [comment]
Appimage is way better than flatpak, you can make it portable very easy by just creating a folder with same name of appimage executable ending with **.home** e.g. TouchHLE-0.2.2-3-anylinux-x86_64.AppImage.home, super useful to move to other computers with your data config saves etc from a usb thumbdrive, super cool!

I like appimage, so as big-endian devices.

Regards,
Link.

**Comment by hikari-no-yume** [comment]
Big-endian? What?
